balancer/base/balancer.go

```diff
@@ -213,21 +213,35 @@ func (b *baseBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.Su
 		}
 		return
 	}
-	b.scStates[sc] = s
-	switch s {
-	case connectivity.Idle:
+	if s == connectivity.Idle {
 		sc.Connect()
-	case connectivity.Shutdown:
+	}
+	if oldS == connectivity.TransientFailure && s != connectivity.Ready && s != connectivity.Shutdown {
+		// Once a subconn enters TRANSIENT_FAILURE, keep it in that state until
+		// it becomes READY in order to prevent the aggregated state from being
+		// always CONNECTING when many backends exist but are all down.
+		// SHUTDOWN is also allowed as that ends the SubConn.
+		s = connectivity.TransientFailure
```

Comment:
Why not make `ConnectivityStateEvaluator` aware of `map[subconn]state` and check the old state? Instead of faking the new state here.

We would need a `ConnectivityStateEvaluator-v2` though, because it's exported.